server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -105,7 +105,7 @@ public MappedFieldType(String name, boolean isIndexed, boolean isStored,
      * for metadata fields, field types should not throw {@link UnsupportedOperationException} since this
      * could cause a search retrieving multiple fields (like "fields": ["*"]) to fail.
      */
-    public abstract ValueFetcher valueFetcher(MapperService mapperService, SearchLookup searchLookup, @Nullable String format);
+    public abstract ValueFetcher valueFetcher(ValueFetcher.Context valueFetcherContext, SearchLookup searchLookup, @Nullable String format);
```

Comment:
Why not using `QueryShardContext` ? We could also remove the `SearchLookup` argument.

Reply:
> Why not using `QueryShardContext` ? We could also remove the `SearchLookup` argument.

I'd prefer not to use `QueryShardContext` everywhere. It really is "too big" as it is.

For what it is worth, we could totally add `SearchLookup` to `ValueFetcher.Context` if we wanted.

But I've made three attempts at this change so far and we didn't like them so I'm kind of done having opinions here. Do whatever makes folks happy.

Reply:
QueryShardContext is a huge object that I don't think we want to spread around more than it already is. While we remove the dependency from MapperService, I think that we should try and be careful with encapsulation and who gets access to what. `sourcePath` is currently not exposed anywhere other than `MapperService`, and we will have to find a way to expose it so we can progress with removing direct MapperService usages, but we want to be careful with adding new methods to all of the existing contexts, that already have many methods.

Reply:
I updated and added QueryShardContext as an argument. I am not sure about SearchLookup because it needs to retrieved specifically through QueryShardContext#newFetchLookup and reused. Looks like we should improve this as its confusing that you can get the lookup also from the context, though it would be the wrong one. I will leave this for another time though.

Reply:
+1 to solve this as a follow up.